scylla-cql/src/types/deserialize/value.rs

```diff
@@ -0,0 +1,1528 @@
+
+        // collections
+        let mut list = BytesMut::new();
+        list.put_i32(3);
+        append_cell(&mut list, &123i32.to_be_bytes());
+        append_cell(&mut list, &456i32.to_be_bytes());
+        append_cell(&mut list, &789i32.to_be_bytes());
+        let list = make_cell(&list);
+        let list_type = ColumnType::List(Box::new(ColumnType::Int));
+        compat_check::<Vec<i32>>(&list_type, list.clone());
+        compat_check::<BTreeSet<i32>>(&list_type, list.clone());
+        compat_check::<HashSet<i32>>(&list_type, list);
+
+        let mut map = BytesMut::new();
+        map.put_i32(3);
+        append_cell(&mut map, &123i32.to_be_bytes());
+        append_cell(&mut map, "quick".as_bytes());
+        append_cell(&mut map, &456i32.to_be_bytes());
+        append_cell(&mut map, "brown".as_bytes());
+        append_cell(&mut map, &789i32.to_be_bytes());
+        append_cell(&mut map, "fox".as_bytes());
+        let map = make_cell(&map);
+        let map_type = ColumnType::Map(Box::new(ColumnType::Int), Box::new(ColumnType::Text));
+        compat_check::<BTreeMap<i32, String>>(&map_type, map.clone());
+        compat_check::<HashMap<i32, String>>(&map_type, map);
+
+        // Tuples
+        let tup_type = ColumnType::Tuple(vec![ColumnType::Text, ColumnType::Int, ColumnType::Uuid]);
+        let mut tup = BytesMut::new();
+        append_cell(&mut tup, "quick brown fox".as_bytes());
+        append_cell(&mut tup, &123i32.to_be_bytes());
+        append_cell(&mut tup, &uuid1.to_u128_le().to_be_bytes());
+        let tup = make_cell(&tup);
+        compat_check::<(String, i32, Uuid)>(&tup_type, tup);
+    }
+
+    // Checks that both new and old serialization framework
+    // produces the same results in this case
+    fn compat_check<T>(typ: &ColumnType, raw: Bytes)
+    where
+        T: for<'f> DeserializeCql<'f>,
+        T: FromCqlVal<Option<CqlValue>>,
+        T: Debug + PartialEq,
+    {
+        let mut slice = raw.as_ref();
+        let mut cell = types::read_bytes_opt(&mut slice).unwrap();
+        let old = T::from_cql(
+            cell.as_mut()
+                .map(|c| deser_cql_value(typ, c))
+                .transpose()
+                .unwrap(),
+        )
+        .unwrap();
+        let new = deserialize::<T>(typ, &raw).unwrap();
+        assert_eq!(old, new);
+    }
+
+    fn deserialize<'frame, T>(typ: &'frame ColumnType, byts: &'frame Bytes) -> Result<T, ParseError>
+    where
+        T: DeserializeCql<'frame>,
+    {
+        <T as DeserializeCql<'frame>>::type_check(typ)?;
+        let mut buf = byts.as_ref();
+        let cell = types::read_bytes_opt(&mut buf)?;
+        let value = cell.map(|cell| FrameSlice::new_subslice(cell, byts));
+        <T as DeserializeCql<'frame>>::deserialize(typ, value)
+    }
+
+    fn make_cell(cell: &[u8]) -> Bytes {
+        let mut b = BytesMut::new();
+        append_cell(&mut b, cell);
+        b.freeze()
+    }
```

Comment:
I'll change it to "make_bytes" here because it's more in line with the protocol spec. The "cell" term is vaguely similar to a concept of "atomic cell" in ScyllaDB's codebase, but it's only superficially similar, so I'll get rid of it here to reduce confusion.